lang/LangPrimSource/PyrUnixPrim.cpp

```diff
@@ -341,6 +342,80 @@ int prGMTime(struct VMGlobals *g, int numArgsPushed)
 	return errNone;
 }
 
+int prDateFromString(struct VMGlobals *g, int numArgsPushed)
+{
+	PyrSlot *dateObjectSlot = g->sp - 2;
+	PyrSlot *timeSlot = g->sp - 1;
+	PyrSlot *formatSlot = g->sp;
+
+	constexpr auto len = 1024;
+	char timeString[len];
+	if (slotStrVal(timeSlot, timeString, len)) return errWrongType;
+	char formatString[len];
+	if (slotStrVal(formatSlot, formatString, len)) return errWrongType;
+
+	using namespace boost::posix_time;
+	auto facet = new time_input_facet(); // will be owned/deleted by the locale
+	facet->format(formatString);
+	std::istringstream ss(timeString);
+	ss.imbue(std::locale(ss.getloc(), facet));
+	ptime pt;
+	if (!(ss >> pt)) {
+		error("time parsing failed\n");
+		return errFailed;
+	}
+	auto tm0 = to_tm(pt);
+	
+	tm0.tm_isdst = -1; // attempt to determine if Daylight Saving Time in effect
+	auto timePoint = std::chrono::system_clock::from_time_t(mktime(&tm0));
+	fillSlotsFromTime(dateObjectSlot, &tm0, timePoint);
+	return errNone;
+}
+
+// Given an incomplete Date structure (missing dayOfWeek and
+// rawSeconds), calculate and fill in those properties.
+int prDateResolve(struct VMGlobals *g, int numArgsPushed)
+{
+	PyrSlot *a = g->sp;
+	PyrSlot *slots = slotRawObject(a)->slots;
+	
+	if (IsNil(slots + 0)) {
+		SetNil(a);
+		return errNone;
+	}
```

Comment:
what is the purpose of this block?

Reply:
This was another attempt at being "conservative".  (-;  If certain of the incoming arguments are Nil, then we won't be able to resolve a date/time (unless we decide on some defaults for that case -- what would be the "default" year/month/day?).  Instead of checking if *any* of the required inputs was Nil (as most typically this is what would happen e.g. if you called `Date.new` with no arguments), I just checked the first argument here (if it's Nil, it sets the function's return value back to scripting as Nil, rather than returning an error...).  But I don't know that this was a good thing to do...what do you suggest for this case?  Return errFailed instead?  Or use some kind of default values?

Reply:
Here's what I would propose.

1. Have default values for the first six arguments (Y,M,D,H,M,S) in `Date.new`, so if you call Date.new it would create *something* valid and not give an error (as it does with my current changes, and did *before* my changes!).  For everything except `year`, there is an obvious and reasonable default (month=1, day=1, hour=0, minute=0, second=0).  For year, we could just arbitrarily choose something.  In the Date class, someone decided to initialize the year member to 2000, although the existing `new` method always overwrites what's there, so there effectively is no default date.  So if you do `Date.new` right now, you get an error, because all the members are set to Nil, and then the asString method fails when the post window tries to output an "empty" Date object.  But the year could arbitrarily default to 2000, or 2017, or have `Date.new` give the current time (`Date.localtime`), or...?  As I mentioned in an earlier comment, I'd really prefer that the `new` method only have Y, M, D, H, M, S arguments, not `dayOfWeek` or `rawSeconds`.  Especially not to have `dayOfWeek`, although I understand the desire to leave it as-is, because it would be a breaking API change.

```supercollider
Date.new;
// Current behaviour: ERROR: Primitive '_ArrayCat' failed.
// Proposed: -> Sat Jan  1 00:00:00 2000 // add defaults for all parameters (e.g. year 2000, or ...?)

Date(2017);
// Current behaviour: ERROR: Primitive '_Date_Resolve' failed. (because month and day don't have default values)
// Proposed: -> Sun Jan  1 00:00:00 2017 // month/day default to 1, hour/min/sec to 0
```

2. If you choose to set `rawSeconds`, it would always override/recalculate all the other values, to remain internally consistent (they are two views on the same data).  Would be true both if you set the `rawSeconds` argument to `new`, but also if you did `d.rawSeconds = <something>` on an existing Date object.

```supercollider
Date(rawSeconds: -14186520);
// Current behaviour: ERROR: Primitive '_ArrayCat' failed.
// Proposed: -> Sun Jul 20 20:18:00 1969 // compute all the other fields from the rawSeconds

// The following would use the rawSeconds value, ignoring the rest of the args since they're inconsistent
Date(2017, 12, 30, 5, 13, 22, rawSeconds: 1234567890);
// Current behaviour (with my changes): -> Sat Dec 30 05:13:22 2017 // ignore rawSeconds
// Proposed: -> Sat Feb 15 00:31:30 2009 // N.B. *not* December 30, 2017! Use rawSeconds, ignore the rest

d = Date(2005, 5, 31);
// Current behaviour: ERROR: Primitive '_Date_Resolve' failed.
// Proposed (implemented in this PR): -> Tue May 31 00:00:00 2005
d.rawSeconds = d.rawSeconds + 10000; // add 10k seconds, which will update the other fields
// Current behaviour: will add 10000 to a valid Date's rawSeconds, but won't modify any other fields (like hour/minute/second, so the Date will now be "inconsistent"
// Proposed: -> Tue Jun  1 01:46:40 2005 // update other fields to match new rawSeconds
```

3. Note that rawSeconds is normally a Double (and will indeed be fractional, if you use methods like `Date.localtime` to get the high-resolution current time).  Seconds is currently always an Int (generally 0-59).  I wouldn't want to modify this, because it would affect the code that outputs Date strings.  Just to be aware, though: if you have a high-precision Date (e.g. from Date.localtime), and then create a new Date using just the computed YMD/HMS values, you won't get an equal Date (because mtkime only resolves seconds).  I could make it add in the factional bit to the rawSeconds, but right now it doesn't...

```supercollider
d = Date.localtime;
[d.asString, d.rawSeconds].postln;
// -> [ Sun Oct 29 18:07:03 2017, 1509296823.8734 ]

e = Date(d.year, d.month, d.day, d.hour, d.minute, d.second);
[e.asString, e.rawSeconds].postln;
// -> [ Sun Oct 29 18:07:03 2017, 1509296823 ]
d == e;
// -> false

// However, if I implement what I propose above, you could do:
e = Date(rawSeconds: d.rawSeconds);
// The you would get
d == e;
// -> true
```

Also, note that `d.second` for the above example (also for `e`) returns `3`, not `3.8734` (which it *might*, but doesn't!).  I don't think I'd like to change this (this is much more likely to break existing code), although there certainly could be an argument made for that behaviour (fractional seconds) to make sense!

4. One last thing.  With my current changes, if you pass the constructor an invalid date (like day-of-month 60, or hour 200), it will "correct" it to the equivalent valid date.  This is quite cool and useful, see my examples in the documentation or unit tests, to use it to calculate relative dates.  However, I don't do any validation (right now) if you modify the properties individually, once you have an existing Date object.  I could, though, always do revalidation.  So, for example, if you did:

```supercollider
d = Date(2005, 5, 31);
// -> Tue May 31 00:00:00 2005

// You could offset a number of months or days (or years, hours, minutes, seconds) in the future or past quite easily:
d.month = d.month + 20;
// -> Wed Jan 31 00:00:00 2007

d.hour = -50; // 50 hours earlier
// -> Sun Jan 28 22:00:00 2007
```

Anyhow, that's it for now...  What do you think?  Should I implement at least the changes in those first two points?  Thanks.

Reply:
These are all essentially changes to behavior of the constructor except for (4), right?

I think I understand where you're going now with this. Let me state the spec for the constructor a little more concisely:

1. If all args nil -> return "now" (or all-defaults, but I personally find that less useful)
2. Else, if rawSeconds is provided, use it to set (YMD:hms) regardless of whether they are nil or not.
3. Else, if year is not provided, throw an error (can't fill in defaults)
4. Else, fill in defaults for any of (MD:hms) that are nil, and set rawSeconds from (YMD:hms).
5. Set dayOfWeek, overwriting any existing value

Is this correct? From my understanding dayOfWeek is always set internally, possibly changing the user-provided value if it's different.

I think you should provide the resolution operation as a user-facing utility. In other words, leave the current behavior for property setters as-is, but also give the ability to call `resolve` to turn it into a valid date again. You might have to provide two - one to resolve from `rawSeconds`, and one to resolve from other fields. Or, have `resolve` always use the other fields, and rely on the constructor for the second case.

I think paying for what you use is more sensible here, since it's fairly expensive to recompute all the fields on every set of minute or second.

```supercollider
d.hour = d.hour - 1;
d.month = d.month + 3;
d.minute = d.minute + 100;
d.resolve;

d.rawSeconds = d.rawSeconds + 3.5;
d.resolveFromRaw; // first idea

d = Date(d.rawSeconds + 3.5); // second idea
```

Re point (3), I think leave it as is. Currently, you can get millis with `d.rawSeconds % 1`, and making `d.seconds` return a rounded value, instead of truncated, would fudge that. And besides as you point out, it would probably not be very kind to suddenly change the contract of such a low-level and subtle feature.

still not convinced we've pinned down the best solution, but this is getting closer I think!

Reply:
That sounds good to me.  For *your* point 3), I guess you mean "if year is nil but some of (MD:hms) are non-nil, throw an error"?  I'm just noting that, because up to that point, I'd been planning to provide the defaults for MD:hms in scripting:

```supercollider
*new { arg year, month=1, day=1, hour=0, minute=0, second=0, dayOfWeek, rawSeconds;
```

But then if you call Date.new you'll get Y:nil and the rest with default values, so your point 1) would be hard to detect (unless the user wrote `Date(nil, nil, nil, nil, nil, nil)`, which would be pretty annoying! ;-).  So I'll have to have no defaults in the `*new` arguments declaration.

Quick summary with code examples (ready for unit tests!):

```supercollider
Date.new; // current local time
// -> Mon Oct 30 20:25:35 2017
Date(rawSeconds: 1234567890); // init from rawSeconds
// -> Sat Feb 15 00:31:30 2009
Date(1990, 1, 1, 0, 0, 0, rawSeconds: 1234567890); // rawSeconds overrides others
// -> Sat Feb 15 00:31:30 2009
Date(month: 3); // year or rawSeconds must be provided
Date(nil, 1, 1, 0, 0, 0) // ditto: year can't be nil
// ERROR: year must be specified
Date(1990); // defaults exist for MD:hms
// -> Mon Jan  1 00:00:00 1990
Date(1990, 2, 1, dayOfWeek: 1); // dayOfWeek (if we keep it as an arg) is always overwritten
// -> Thu Feb  1 00:00:00 1990
Date(dayOfWeek: 1, rawSeconds: 633826800); // same if you init using rawSeconds, dayOfWeek overwritten
// -> Thu Feb  1 00:00:00 1990
```

Just to be clear on another point... you're still convinced I should keep the Date.new signature the same, with dayOfWeek, even if it's totally useless (just to maintain "backward compatibility")?

Thanks again for this back and forth, it's very helpful!

Reply:
BTW, I'm fine with your `resolve` and `resolveFromRaw` suggestion; I'd thought of maybe doing something similar, that maybe it was too "heavy" to resolve all the time -- and this also allows several values to be modified before calculating the "resolve", as you say.

Reply:
> you're still convinced I should keep the Date.new signature the same, with dayOfWeek, even if it's totally useless (just to maintain "backward compatibility")?

Yes, it's not really that bad to keep it in, although i understand your annoyance :)

> BTW, I'm fine with your resolve and resolveFromRaw suggestion; I'd thought of maybe doing something similar, that maybe it was too "heavy" to resolve all the time

++ what do you think about leaving out `resolveFromRaw` in favor of using the constructor? I think it would be about the same performance-wise and I don't see much of a reason to add a second way to do it beyond it potentially being faster.

> I guess you mean "if year is nil but some of (MD:hms) are non-nil, throw an error"?

Yes -- it's in an if-else branch, the first `if` of which is taken if all of (YMD:hms) are nil. So we only get to point (3) if one of those is non-nil. Sorry if that was confusing

> Thanks again for this back and forth, it's very helpful!

Same to you :)

Reply:
> ++ what do you think about leaving out resolveFromRaw in favor of using the constructor? I think it would be about the same performance-wise and I don't see much of a reason to add a second way to do it beyond it potentially being faster.

That's fine, totally agree, will do that.

So I'll go ahead and implement the last of these details we've talked about...

My last remaining "unknown", then, is about the unit testing stuff.  I'll write my unit tests.  But about the Travis automation -- any resolution/clarity on that?  Or should I just take out the bit I added (hte lines in `travis_test_run_proto.json`)?  If not, I'll need some guidance on how to enable Travis testing properly, and whether I'm on the right track by adding it to the (deprecated) CommonTests repo...or not... (I understand that if so, it's a temporary solution)

Thanks.

Reply:
> But about the Travis automation -- any resolution/clarity on that? Or should I just take out the bit I added (hte lines in travis_test_run_proto.json)? If not, I'll need some guidance on how to enable Travis testing properly, and whether I'm on the right track by adding it to the (deprecated) CommonTests repo...or not... (I understand that if so, it's a temporary solution)

For the sake of us all staying healthy, I think let's just go with the quick way for now - you've done it right with the lines you've added here, and adding it to CommonTests is the way to go. But yes, hopefully very temporary. Sorry again for the confusion.